File: app/src/main/java/com/alexkn/syntact/restservice/SyntactService.java
package com.alexkn.syntact.restservice;

import java.util.List;

import okhttp3.ResponseBody;
import retrofit2.Call;
import retrofit2.http.Body;
import retrofit2.http.GET;
import retrofit2.http.Header;
import retrofit2.http.POST;
import retrofit2.http.Path;
import retrofit2.http.Query;
import retrofit2.http.Url;

public interface SyntactService {

    @GET("phrases/")
    Call<List<Phrase>> getPhrases(@Header("Authorization") String token);

    @GET("templates/")
    Call<List<Template>> getTemplates(@Header("Authorization") String token);

    @GET("templates/{id}/phrases/")
    Call<List<Phrase>> getPhrases(@Header("Authorization") String token, @Path("id") Long templateId, @Query("minid") long minid,
            @Query("limit") int limit);

    @GET("phrases/{id}/translations/")
    Call<List<Translation>> getTranslations(@Header("Authorization") String token, @Path("id") Long phraseId, @Query("lang") String lang);

    @GET
    Call<List<Translation>> getTranslations(@Header("Authorization") String token, @Url String url, @Query("lang") String lang);

    @GET
    Call<List<Phrase>> getPhrases(@Header("Authorization") String token, @Url String url, @Query("minid") long minid, @Query("limit") int limit);

    @POST("templates/{id}/phrases/")
    Call<ResponseBody> postPhrases(@Header("Authorization") String token, @Path("id") Long templateId, @Body List<PhrasesRequest> phrasesRequests);

    @POST("templates/")
    Call<Template> postTemplate(@Header("Authorization") String token, @Body TemplateRequest templateRequest);
}

File: app/src/main/java/com/alexkn/syntact/restservice/SyntactService.kt
package com.alexkn.syntact.restservice

import okhttp3.ResponseBody
import retrofit2.Call
import retrofit2.http.*

interface SyntactService {

    @GET("phrases/")
    fun getPhrases(@Header("Authorization") token: String): Call<List<Phrase>>

    @GET("templates/")
    suspend fun getTemplates(@Header("Authorization") token: String): List<Template>

    @GET("templates/{id}/phrases/")
    fun getPhrases(@Header("Authorization") token: String, @Path("id") templateId: Long?, @Query("minid") minid: Long,
                   @Query("limit") limit: Int): Call<List<Phrase>>

    @GET("phrases/{id}/translations/")
    fun getTranslations(@Header("Authorization") token: String, @Path("id") phraseId: Long?, @Query("lang") lang: String): Call<List<Translation>>

    @GET
    suspend fun getTranslations(@Header("Authorization") token: String, @Url url: String, @Query("lang") lang: String): List<Translation>

    @GET
    suspend fun getPhrases(@Header("Authorization") token: String, @Url url: String, @Query("minid") minid: Long, @Query("limit") limit: Int): List<Phrase>

    @POST("templates/{id}/phrases/")
    fun postPhrases(@Header("Authorization") token: String, @Path("id") templateId: Long?, @Query("lang") lang: String,
                    @Body phrasesRequests: List<PhrasesRequest>): Call<ResponseBody>

    @POST("templates/")
    fun postTemplate(@Header("Authorization") token: String, @Body templateRequest: TemplateRequest): Call<Template>

    @DELETE("templates/{id}")
    fun deleteTemplate(@Header("Authorization") token: String, @Path("id") templateId: Long?): Call<ResponseBody>
}
